client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoResultSet.java

```diff
@@ -144,27 +145,39 @@ private static <T> Iterator<T> flatten(Iterator<Iterable<T>> iterator, long maxR
         return stream.iterator();
     }
 
-    private static class AsyncIterator<T>
+    @VisibleForTesting
+    static class AsyncIterator<T>
             extends AbstractIterator<T>
     {
         private static final int MAX_QUEUED_ROWS = 50_000;
         private static final ExecutorService executorService = newCachedThreadPool(
                 new ThreadFactoryBuilder().setNameFormat("Trino JDBC worker-%s").setDaemon(true).build());
 
         private final StatementClient client;
-        private final BlockingQueue<T> rowQueue = new ArrayBlockingQueue<>(MAX_QUEUED_ROWS);
+        private final BlockingQueue<T> rowQueue;
         // Semaphore to indicate that some data is ready.
         // Each permit represents a row of data (or that the underlying iterator is exhausted).
         private final Semaphore semaphore = new Semaphore(0);
-        private final CompletableFuture<Void> future;
+        private final Future<?> future;
+        private volatile boolean cancelled;
+        private volatile boolean finished;
```

Comment:
is the `finished` field for testing only?
why not check whether future.isDone() instead, and not have this field?

Reply:
Yes, the problem is that threads may leak in current implementation even though future is cancelled (`future.isDone() == true`), so I add this test to make sure thread goes to the end of its work.